src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -278,31 +281,99 @@ def process_input(conf):
+            os.link(cfn, dst)
+
+            metafile = os.path.join(dstd, f"{upa}.meta")
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
 
-        # cfn points to the assembly file outside of the container
-        # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
-        # workdir is shared between the container and the external file system
-        # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+        if retrieve_sample:
+            _download_sample_data(conf, upa)
 
-        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
-        # upa file is downloaded to cfn
-        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
 
-        # each upa in output_dir as a seperate directory
-        dstd = os.path.join(conf.pth, upa)
-        os.makedirs(dstd, exist_ok=True)
+def _download_sample_data(conf, upa):
+    # retrieve sample data from sample service and save to file
 
-        dst = os.path.join(dstd, f"{upa}.fa")
-        # Hard link .fa file from job_dir to output_dir in WS
-        os.link(cfn, dst)
+    dstd = os.path.join(conf.pth, upa)
+    os.makedirs(dstd, exist_ok=True)
+    sample_file = os.path.join(dstd, f"{upa}.sample")
 
-        metafile = os.path.join(dstd, f"{upa}.meta")
-        # save meta file with relevant object_info
-        with open(metafile, "w", encoding="utf8") as json_file:
-            json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+    if os.path.isfile(sample_file):
+        print(f"Skip downloading sample for {upa} as it already exists")
+        return
 
-        print("Completed %s" % (upa))
+    # retrieve data links associated with upa
+    links_ret = _post_sample_service(conf.token,
+                                     conf.sample_url,
+                                     "get_data_links_from_data",
+                                     {"upa": upa.replace("_", "/")})
+    data_links = links_ret['links']
+    if not data_links:
+        print(f"No sample data links found for {upa}")
+        return
+
+    # there should only be one data link for each upa
+    if len(data_links) != 1:
+        raise ValueError(f"Expected 1 data link for {upa}, got {len(data_links)}")
+
+    # retrieve sample data and save to file
+    sample_id = data_links[0]['id']
+    sample_ret = _post_sample_service(conf.token,
+                                      conf.sample_url,
+                                      "get_sample",
+                                      {"id": sample_id})
```

Comment:
```suggestion
                                      "get_sample_via_data",
                                      {"upa": upa, "id": sample_id, "version": data_links[0]["version"]})
```

No guarantee we'll have direct access to the sample

Reply:
👍

Reply:
:+1: